compiler/rustc_index/src/bit_set.rs

```diff
@@ -328,6 +342,403 @@ impl<T: Idx> BitRelations<BitSet<T>> for BitSet<T> {
+            // the final one.
+            let final_chunk_domain_size = {
+                let n = domain_size.index() % CHUNK_BITS;
+                if n == 0 { CHUNK_BITS } else { n }
+            };
+            let chunks = iter::repeat(CHUNK_BITS)
+                .take(num_chunks(domain_size) - 1)
+                .chain(iter::once(final_chunk_domain_size))
+                .map(|chunk_domain_size| Chunk::new(chunk_domain_size, is_empty))
+                .collect();
+            let res = ChunkedBitSet { domain_size, chunks, marker: PhantomData };
+            res.sane();
+            res
+        }
+    }
+
+    /// Creates a new, empty bitset with a given `domain_size`.
+    #[inline]
+    pub fn new_empty(domain_size: usize) -> Self {
+        ChunkedBitSet::new(domain_size, /* is_empty */ true)
+    }
+
+    /// Creates a new, filled bitset with a given `domain_size`.
+    #[inline]
+    pub fn new_filled(domain_size: usize) -> Self {
+        ChunkedBitSet::new(domain_size, /* is_empty */ false)
+    }
+
+    #[cfg(test)]
+    fn chunks(&self) -> &[Chunk] {
+        self.sane();
+        &self.chunks
+    }
+
+    /// Returns `true` if `self` contains `elem`.
+    #[inline]
+    pub fn contains(&self, elem: T) -> bool {
+        self.sane();
+        assert!(elem.index() < self.domain_size);
+        let chunk = &self.chunks[chunk_index(elem)];
+        match &chunk {
+            Zeros(_) => false,
+            Ones(_) => true,
+            Mixed(_, _, words) => {
+                let (word_index, mask) = chunk_word_index_and_mask(elem);
+                (words[word_index] & mask) != 0
+            }
+        }
+    }
+
+    /// Insert `elem`. Returns whether the set has changed.
+    pub fn insert(&mut self, elem: T) -> bool {
+        self.sane();
+        assert!(elem.index() < self.domain_size);
+        let chunk_index = chunk_index(elem);
+        let chunk = &mut self.chunks[chunk_index];
+        match *chunk {
+            Zeros(chunk_domain_size) => {
+                if chunk_domain_size > 1 {
+                    let mut words = [0; CHUNK_WORDS];
+                    let (word_index, mask) = chunk_word_index_and_mask(elem);
+                    words[word_index] |= mask;
+                    *chunk = Mixed(chunk_domain_size, 1, Rc::new(words));
+                } else {
+                    *chunk = Ones(chunk_domain_size);
+                }
+                self.sane();
+                true
+            }
+            Ones(_) => false,
+            Mixed(chunk_domain_size, ref mut count, ref mut words) => {
+                // We skip all the work if the bit is already set.
+                let (word_index, mask) = chunk_word_index_and_mask(elem);
+                let changed = if (words[word_index] & mask) == 0 {
+                    let words = Rc::make_mut(words);
+                    words[word_index] |= mask;
+                    *count += 1;
+                    if *count == chunk_domain_size {
```

Comment:
The point about the check is good, I have addressed that in `insert` and `remove`.

A global Rc makes sense for Zeros because it's always all zeros. But Ones chunks can have trailing zeroes (and several operations rely on these) so you'd only be able to have a global Rc for the case where the ones fill all the way to the end.

But I don't think the global Rc would help anyway? As soon as you move away from all zeroes or all ones you have to duplicate anyway.